### ISSUE TITLE
[Suggestion?] Power Balancing and Resource Management

### ISSUE THREAD

**Issue by VermillionX** [BluSunrize/ImmersiveEngineering#505]
**Disclaimer:** For instructive purposes I will be referencing data concerning primary mainstream tech mods including the "Holy Trinity" (Buildcraft, Forestry and Railcraft.) and the current leading generic tech mod: Thermal Expansion.

**Caution: Wall of text ahead!**

**Power Production:**
In regards to resource cost, the Kinetic Dynamo has a vast discrepancy between being fitted with a Water Wheel in comparison to a Wind Mill. A Windmill is reliant on both height, free space and weather. A fair trade for unlimited power without the fuel cost of a powered engine/dynamo. The higher you place your Kinetic Dynamo, the longer the cable has to be to reach your base on the ground (unless you limit yourself by building in the air), meaning greater energy loss (which is around 50% of the original output btw).
_Note: A Windmill, at max height produces 22 RF/t with a basic blade and 44 with the advanced version (30/60 RF/t during a storm). Just above the ground level, it produces around 6 RF/t, but won't suffer as much cable loss._
So to me, a Windmill is pretty fair. If you want it to work, you have to put work and resources in.

Now comes the Waterwheel: It's not dependant on anything except fluid pouring over it. If you know what you're doing, you can get it to output 110+ RF/t using 3 wheels. Strangely enough, of all the people who've tried to find the optimal output of the kinetic dynamo, they weren't overly imaginative. You can max out the output by switching out water for much denser Lava (Waterwheel with lava = Magmawheel?), and getting an extra 9-10 RF/t by applying a stream of Zephyrean Aerotheum (Requires TE) to the back side (it flows upwards). Feel free to try other fluids, but I found Creosote oil to be almost as good as lava.
Anyway, as you can see there's an absolutely massive gap between the more expensive and restrictive Windmill (Advanced Windmill also requires steel) and the cheaper Waterwheel (Just treated wood. 20 planks or so.)
I suggest greatly reducing the power output of the waterwheels from 30-35 RF/t per wheel to 6-10 RF/t per wheel (6 water, 10 lava). The power output after Windmill cable losses would make it about even with the Waterwheels, who don't suffer from requiring a long cable. Whereas the player can gain bonus power for a windmill from a storm, they can also gain addition waterwheel power by risking the use of lava instead of water.

Finally, the Thermoelectric Generator: Assuming you're not using TE you'll be stuck using the basic lava and water or packed ice combo (15-16 RF/t); otherwise the Pyrotheum and Cryotheum combo gives 31RF/t.
The ThermoGen is around the same cost as the Kinetic Dynamo to craft, isn't affected by the enviroment (aside from the hot/cold fluids you're using) and like the dynamo outputs energy infinitely. Unlike the Kinetic Dynamo however, it doesn't take up much space and you can run multiple ThermoGens from a single pool of lava/pyro/cryo. It doesn't take much to just make 6 or so early on and just hide them in your floor and never need to think about them ever again.
I suggest a minimal change: a very low chance to "evaporate" (delete) their cold fluid or cool (into cobble) their hot fluid. Meaning players will need to check in on them every now and then.
A small inconvenience every few hours for fuel-less power doesn't sound too bad. Just like Solar power doesn't work at night. Of course, I expect some people to display annoyance at the threat of needing to maintain their machines or dare I say... play the game?

Naturally, as a course of requiring more generators or better setups, IE's power storage comes more to the fore. At present, the IE power system stores a fraction of TE's cheaper energy cells and is less user-friendly, requiring a Voltmeter to even check its levels (unless you have waila installed). With it being harder to sustain enough power to run multiple machines constantly,

**Power Consumption:**
Fairly minor, but here come the comparisons.
Industrial Fermenter: Currently consumes 10 RF/t. That's so small as to be run from a single windmill. It's equivalent would be the Forestry Fermenter (170 RF/t), however since the Forestry Fermenter's outputs are usable as a fuel and IE's are single-use in a Refinery a difference in power is expected. I suggest increasing the power requirement to 32-40 RF/t. It does afterall, run at least 4x faster than the Forestry Fermenter.
Industrial Squeezer: Also consumes 10 RF/t and likewise I suggest the same increase in consumption rate of power.
Crusher: Currently consumes 80 RF/t. While moderately expensive to build, it's not hard to get steel with IE's Blast Furnace being cheaper than RailCraft's, but lacking the ability to break down metal tools, armor and blocks into steel. Anyway, the Crusher operates at a little over 4x Thermal Expansion's basic Pulverizer's speed at double the RF cost. However, with a Reinforced Pulverizer with Speed upgrades to match the Crusher, the power consumption of the Pulverizer maxes out at 320 RF/t, while leaving space for more upgrades to increase secondary yield.
As such, I suggest the Crusher's max power consumption be buffed to 120 RF/t. This is a compromise with TE by being faster, it's also more expensive, larger, cannot be upgraded and lacks an all-important GUI unlike the Pulverizer. So it can still run on lower voltages, as it does now, it just takes a little more power to get it up and running at maximum efficiency.
Refinery: Currently uses 80 RF/t and I think that's just fine with the boost to the Squeezer and Fermenter. Functionally and realistically, Biodiesel is almost identical in burn rate as Forestry Biofuel but the efficiency of the Fermenter and Squeezer (IE's ones, not Forestry's) gave it a 10% or so power bonus return. They'd sit almost even now.
As for the External Heater, Arc Furnace and Excavator I think they're fine. Though personally, I think the Excavator should output cobble, gravel, sand, dirt and clay as a bonus to accompany a much faster bucketwheel.(e.g. the wheel turns 5 times as opposed to the 1 it does now; on the 5th turn it gets ores, but the extra 4 turns always pull "dirt" items).

**Power Management Summary:**
With the current rates, the basic generators, no matter how poorly set up can run the low-mid tier machines effortlessly. If all/most of the machines consume less power than a basic generator or two provides, the power storage Capacitors are nothing more than buffers for the end-game machines when the players eventually get to them instead of tiered storage for each level of machine. Even then, only the HV Capacitor would be relevant to an Arc Furnace or Excavator (as if you'd take an HV capacitor to run an Excavator).
Unless players have a reason to actively build power storage for practical use, well, they won't. By boosting the low tier machines to consume more, players will need to learn to effectively manage their power early-game and rely on energy storage instead of meeting all their needs with only one dynamo. As they progress and make more generators and redundant storage they can move from running one machine some of the time to running all of their machines most of the time, to eventually being able to run everything effortlessly. With the Capacitors of every tier being used as storage buffers for all machines instead of just an Arc Furnace.
With the changed generator values, this would be rather time consuming if IE was used alone; but since it's predominately used with either Thermal Expansion or BuildCraft, they can easily take over the mid-tier power generation with their fuelled power generators, which IE currently lack (Excluding the end-game Diesel Gen).

**Resource Management:** (Doesn't really matter)
Fairly simple, it's something that BC, Forestry, RC and IC2 all managed to do years ago; and that was generate resources in-world that match the consumption rate of their machines and is something that really every mod should be aspiring for. If there's too many ores in the ground, people end up with an excess which prevents demand on a server.
Immersive Engineering is excellent with Iron Consumption, using large quantities of iron for many things. However, Aluminium, Nickel, Silver and Lead all generate in large quantities but are severely underused in the amounts they generate in.
Though really, this is only relevant if you're not using Thermal Foundation for all your oregen needs. Where Thermal Foundation just increases the amount of ores that generate every time another mod becomes commonly used so you can just make every machine instead of determining which one you need most and rationing your resources.
My suggestion for IE's oregen is to remove both silver and lead ores from the world, leaving them as excavator veins and replacing them in-world with Galena Ore (50% silver, 50% lead); Where Galena Ore, with the spawn ratio of the current Lead Ore, can be smelted normally for 1 Silver, or processed with a Crusher, Arc Furnace, Pulverizer, etc... for 1 silver and 1 lead.

**tl;dr**
Suggestions:
- Reduce the power of Waterwheels to 6-10 RF/t per wheel. They produce more than an upgraded mid-tier TE Dynamo atm (110+ RF/t).
- Add a tiny chance for thermogens to turn their lava/pyro into cobble or delete their water/cryo block. Otherwise, suggest halving their output.
- Increase the power consumption of both squeezer and fermenter to 32-40 RF/t and the Crusher to 120 RF/t to promote better early-game power management, instead of no power management.
- External Heater, Arc Furnace and Excavator are fine.
- Suggest removing (useless) Lead and Silver Ores from worldgen and replacing with Galena Ore. There's no point in having dozens of an ore in each chunk and having little to no use for the amount you get.

_Note: If you have Thermal Expansion installed, the above changes probably wouldn't make much of a difference to you. Some things would consume more power, but you'd probably be using TE's machines anyway._

**Comment by mindforger** [comment]
i must agree, that with one kin gen + 3 water wheels i achived an output of 77RF/t, where the windmill does only provide around 10RF/t "closer" to ground ... like when used in a semi-decorative manner it is pretty useless, something like 45RF/t for a full blown watermill and maybe 15 for a normal and 30RF/t for an advanced windmill sound fair (at height 90, to maintain visiblility from ground)

i disagree hereby with the "all-important GUI" i like the crusher beeing what it is, a crusher and not some fancy interface controlled machine but thats really just my opinion
the consumption rate is all in all okay, because you can not modify it's output chances and it has a fair speed

i would appriciate it a lot when there could be an option, instead of an "empty shovel" chance, the excavator would output something like dirt/gravel/cobble/sand instead any ore then ... 
(i nerved it in our world to 50% speed and 15% yield with 15% empty shovels, because it was insane how much it produced in a very short time)
imho, leave the default as it is, but keep it configurable ... i know there have been a lot of cahnges in calculation and ore deposit configuration but i was honestly too lazy to build myself the latest few versions to dig deeper into it

consumptionrate(per tick) for the fermenter and squeezer could possibly be a little bit higher by default, but be configurable does also do the job ... imho

removing ores from the world is savegame breaking! simply a no-go if you ask me ... but blusunrize did already deal with a "somethign needs to be replaced" situation

**Comment by VermillionX** [comment]
The Free-power generators really shouldn't be higher than any fuelled generator. Where a fuelled generator has a continuous resource cost in exchange for higher power output, the Kinetic/Thermo generators consume nothing and run 24/7, unlike the standard Solar system most mods use that only work during the day. 45RF/t for a single kinetic dynamo is massive compared to the consumption of IE's machines, even with boosted consumption rates. That's over half a TE Dynamo, which happens to also be equal to a BC/RC endgame engine.

Also, I didn't say the Crusher should have a GUI, I said that it's a trade-off against the Pulverizer that it shouldn't have one. That in exchange for consuming 120 RF/t, over an upgraded Pulverizer at 320 RF/t, it makes up the difference in its inflexibility and construction cost.

I used Lodeloader to add custom excavator veins to my own test worlds that include cobble, stone, sand and whatnot alongside standard ores and nuggets. A pity dirt and gravel don't work with it and that you can't disable/overwrite existing veins.

As for Galena, he could simply overwrite the Lead ore with Galena and remove the silver ore from the future worldgen. The Ores would still exist, it just wouldn't generate anymore and the lead ores would be replaced by the better-balanced Galena.

Really in my opinion, Immersive Engineering shouldn't be used with Thermal Expansion. They have too many competing and overlapping features, with differing power balancement. It's better used with BuildCraft, Railcraft and Forestry (IC2 also fits in with these three by design).
- Where TE's power transfer is lossless, both IE's and BC's are lossful.
- Where TE's item transport is instantaneous, both IE's, BC's and Railcraft all work using physical movement from one point to another.
- Where TE's power storage is magnitudes higher than IE's, BC's is almost non-existant (Giving a nice functional energy storage niche for IE to slot itself into due to the difference in scale).
- Where TE adds massive amounts of resources and uses little of them, BC/RC/FR balance themselves with what they add (Example: A forestry Autofarm costs around 120 copper and tin).
- Where TE adds one-block portable item and fluid storage blocks (With massive storage capacity), IE and RC both add large expensive multiblock tanks.
- Then there's TE's machines... Pulverizer overlaps/competes with the Crusher; Induction Smelter competes with the Arc Furnace; Redstone Furnace competes with the External Heater; Cyclic Assembler (will) competes with the Assembler; Portable Tanks compete with the new IE Fluid Tank whose name I don't know and Caches compete with the Silos.

In almost every case, TE comes out on top when compared with IE; So why bother trying to make it work with a mod that doesn't want you to?
